Product Title block: add support global style

gigitux opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

This PR adds support for global style for the `Product Title` block.

Now, the user can edit the style for:
- line-height
- font family
- font-weight
- text transform
- background-color
- margin
- font-size

<!-- Reference any related issues or PRs here -->
Fixes #4965 

### Screenshots

![image](https://user-images.githubusercontent.com/4463174/148072903-c0553e9b-95a5-489f-9a62-f8044488d840.png)
*without global style*


![image](https://user-images.githubusercontent.com/4463174/148072776-1e9b69e2-b074-4c57-8643-8c64af3746b5.png)
*with global style*


### Testing

### Manual Testing

Check out this branch:

1. Install and enable the `Gutenberg` plugin on WordPress 5.9 site.
2. Install and enable the `Twenty Twenty-Two` theme.
3. Add the `Single Product block` (this block contains `Product Title Block`) to a post.
4. Get the focus on `Product Title block`.
5. On the right sidebar, personalize the styles of the block.
6. Go on the page and check if there are changes.
7. Reset to default using the `Reset` button from the different sections. 
8. Go to Dashboard and select Appearance > Editor (beta).
9. On the Editor page click on the `Styles` icon on the right-top corner.
10. Verify that the `Product Title block` is shown under the `Blocks` section. Personalize again the block.
11. Save your changes.
12. Go on the page created earlier and check if all styles are applied correctly.
13. Edit your previous post/page again.
14. Change again the styles.
15. Save your changes.
16. Check if these styles have priority over the styles from the Site Editor.

### Changelog

> Add support for the global style for the Product Title block.

gigitux committed:
Product title: add support global style #4965

gigitux committed:
add specific type

dinhtungdu commented on assets/js/atomic/blocks/product-elements/title/index.ts:
```diff
@@ -24,17 +24,25 @@ const blockConfig: BlockConfiguration = {
+			} ),
+		},
+		color: {
+			text: false,
```
Can you explain why text color is disabled here? Given that we allow changing the background, it makes more sense to me to allow changing text color too.

gigitux replied:
> The block style is working for me 🎉 , both color and typography are working. But for the global style, the background color didn't work in my test. Please check the screencast for me detail.
> 
>  Screen.Recording.2022-01-05.at.16.49.06.mov

Mmm, it seems that it is a GB problem. What version are you using?

gigitux replied:
By #4965, the color for the text should be disabled, but I agree with you. 
I added the support for text color, too, now I will add a comment on the main issue.

dinhtungdu commented on assets/js/utils/style-attributes-utils.ts:
```diff
@@ -0,0 +1,103 @@
+): Record< string, unknown > => {
+	const keys = [ 'margin' ];
+
+	const getValueOrDefault = ( value: unknown ) => {
+		return isString( value ) && value.length > 0 ? value : '0';
+	};
```
I don't think we should return `0` as the default value here, it should be `undefined` instead. The element may have spacing rules defined in CSS, using `0` will override its styling.

dinhtungdu commented on assets/js/utils/style-attributes-utils.ts:
```diff
@@ -0,0 +1,103 @@
+			: {};
+
+		if ( keys.includes( key ) ) {
+			return {
+				...acc,
+				[ key ]: `${ getValueOrDefault(
+					spacingProperty.top
+				) } ${ getValueOrDefault(
+					spacingProperty.right
+				) } ${ getValueOrDefault(
+					spacingProperty.bottom
+				) } ${ getValueOrDefault( spacingProperty.left ) }`,
+			};
```
Same as above, we should only use shorthand when all four sides are set, otherwise, we should use single side rules.

gigitux replied:
Thanks for this! I updated the function!

dinhtungdu reviewed (commented):
The block style is working for me 🎉 , both color and typography are working. But for the global style, the background color didn't work in my test. Please check the screencast for me detail.

https://user-images.githubusercontent.com/5423135/148203207-059f13e3-c74c-42a5-b569-d091ee277950.mov

gigitux commented:
> The block style is working for me 🎉 , both color and typography are working. But for the global style, the background color didn't work in my test. Please check the screencast for me detail.
> 
>  Screen.Recording.2022-01-05.at.16.49.06.mov

Mmm, it seems that it is a GB problem. What version are you using?

dinhtungdu commented:
> Mmm, it seems that it is a GB problem. What version are you using?

@gigitux I used the [latest `trunk`](https://github.com/WordPress/gutenberg/tree/744c61a1fc57d80bc6630ca44f6c217ef43fa638) (I've just tested again).

gigitux committed:
add custom save function

gigitux commented:
> > Mmm, it seems that it is a GB problem. What version are you using?
> 
> @gigitux I used the [latest `trunk`](https://github.com/WordPress/gutenberg/tree/744c61a1fc57d80bc6630ca44f6c217ef43fa638) (I've just tested again).

Mmm, it is a very strange problem. 
https://user-images.githubusercontent.com/4463174/148233284-f73cd082-0da4-4fc1-b5ee-64e74c1ddd7a.mp4
If you see my video, I don't have this problem. Same with TT1 theme.

I updated the PR with your feedback. I also modified the CSS, because before I hadn't edited the save function to be compliant with apiVersion 2: now everything should be ok 👍

dinhtungdu commented:
> Mmm, it is a very strange problem.
> [user-images.githubusercontent.com/4463174/148233284-f73cd082-0da4-4fc1-b5ee-64e74c1ddd7a.mp4](https://user-images.githubusercontent.com/4463174/148233284-f73cd082-0da4-4fc1-b5ee-64e74c1ddd7a.mp4)
> If you see my video, I don't have this problem. Same with TT1 theme.

@gigitux Which WP version are you using?

gigitux commented:
> @gigitux Which WP version are you using?

WordPress 5.9 - beta 4. What about you?

dinhtungdu commented on assets/js/utils/style-attributes-utils.ts:
```diff
@@ -0,0 +1,123 @@
+		if ( keys.includes( key ) ) {
+			return {
+				...acc,
+				...getSpacingStyleInline( key, {
+					top: getValueOrDefault( spacingProperty.top ),
+					right: getValueOrDefault( spacingProperty.right ),
+					bottom: getValueOrDefault( spacingProperty.bottom ),
+					left: getValueOrDefault( spacingProperty.left ),
+				} ),
```
```suggestion
				...getSpacingStyleInline( key, spacingProperty ),
```

dinhtungdu replied:
I think we can pass spacingProperty as the second argument then handle sides there instead. It makes the callback look cleaner.

gigitux replied:
> I think we can pass spacingProperty as the second argument then handle sides there instead. It makes the callback look cleaner.

It can be an idea, but to be honest I prefer to create a more strict typed code. 

![image](https://user-images.githubusercontent.com/4463174/148389640-401671ed-d157-4b08-bdb3-91f1ed8e87fe.png)


With this code (so without type), we lose some information about how the function `getSpacingStyleInline` works.
I mean without an object with `top`, `right`, `bottom`, and `left` key, it doesn't make sense invoke getSpacingStyleInline` function. With the type `Record<string, unknown>` we lose this information and the developer is forced to read the whole function to understand what kind of data to pass.

But this is just my opinion, if you think that it is better to change it, I will do it :D

dinhtungdu replied:
@gigitux With your example, I don't see the need for `getSpacingStyleInline`, we can add the inline style directly like this. Note that we can use `undefined` properties just fine because they will get omitted during render.

```
		if ( keys.includes( key ) ) {
			return {
				...acc,
				[ `${ key }Left` ]: spacingProperty.left,
				[ `${ key }Right` ]: spacingProperty.right,
				[ `${ key }Top` ]: spacingProperty.top,
				[ `${ key }Bottom` ]: spacingProperty.bottom,
			};
		}
```

dinhtungdu commented on assets/js/utils/style-attributes-utils.ts:
```diff
@@ -0,0 +1,123 @@
+	}
+) => {
+	return {
+		...( isString( values.top ) && { [ `${ key }Top` ]: values.top } ),
+		...( isString( values.right ) && {
+			[ `${ key }Right` ]: values.right,
+		} ),
+		...( isString( values.bottom ) && {
+			[ `${ key }Bottom` ]: values.bottom,
+		} ),
+		...( isString( values.left ) && { [ `${ key }Left` ]: values.left } ),
```
Can we use key: value here then remove the null properties? We can even return style object with null properties because they will be omitted during render. Using spread here is quite hard to read IMO.

dinhtungdu commented on assets/js/utils/style-attributes-utils.ts:
```diff
@@ -0,0 +1,123 @@
+	};
+};
+
+export const useColorProps = ( attributes: unknown ): WithStyle & WithClass => {
```
Can we use the correct type of `attributes` here instead of `unknow`? I don't see we're passing string to this hook anywhere. By doing so we don't need this line:

```
const attributesObject = isObject( attributes ) ? attributes : {};
```

gigitux replied:
The problem is that this code is also used on javascript code (so without types), so the check should be done on the js side. I preferred to do it here, so as not to repeat the same check in all the blocks.

dinhtungdu reviewed (commented):
@gigitux With a fresh installation, the global style works as expected now. I left some comments about the utilities, please take a look.

dinhtungdu commented:
@gigitux Looking at https://github.com/WordPress/gutenberg/blob/0666dee9235aedd3d1fa51c733d0c1e146227452/packages/block-library/src/button/edit.js#L26, I'm wondering if we can reuse the spacing handler in Gutenberg instead of creating our own utilities? Like what we do with `useColorProps`.

Edit: By the way, we're creating custom hooks, so `style-attributes-utils.ts` may not be the best place for those hooks.

gigitux committed:
move hooks in a specific folder

gigitux commented:
> @gigitux Looking at https://github.com/WordPress/gutenberg/blob/0666dee9235aedd3d1fa51c733d0c1e146227452/packages/block-library/src/button/edit.js#L26, I'm wondering if we can reuse the spacing handler in Gutenberg instead of creating our own utilities? Like what we do with `useColorProps`.
> 
> Edit: By the way, we're creating custom hooks, so `style-attributes-utils.ts` may not be the best place for those hooks.

Thanks for pointing this out. With my first experiments, it returned to me an empty object. Now I have managed to get it to work correctly.

I moved the hooks under the path `/js/style-attributes.ts`. Let me know if what do you think 👍 

Update: Unfortunately the API `__experimentalGetSpacingClassesAndStyles` is not available on WP 5.8. What should we do? Should we maintain the previous code for WordPress 5.8 and use the new hook for WordPress 5.9?

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/4965-product-title

dinhtungdu reviewed (approved):
@gigitux Great work on this PR 👏🏽 !

> Update: Unfortunately the API __experimentalGetSpacingClassesAndStyles is not available on WP 5.8. What should we do? Should we maintain the previous code for WordPress 5.8 and use the new hook for WordPress 5.9?

As long as our custom hook is feature-gated and no fatal error throws on WP 5.8, I think we can just focus on 5.9, it's not worth the effort to support 5.8 given that most of the global style functionalities are available from 5.9. cc @Aljullu for more insights : ).

gigitux commented:
> As long as our custom hook is feature-gated and no fatal error throws on WP 5.8, I think we can just focus on 5.9, it's not worth the effort to support 5.8 given that most of the global style functionalities are available from 5.9. cc @Aljullu for more insights : ).

Mmm, we have got a fatal error on WP 5.8 (without GB)
![image](https://user-images.githubusercontent.com/4463174/148518343-85532aca-408e-45c2-86de-e2647deb556b.png)

The error:
![image](https://user-images.githubusercontent.com/4463174/148518642-da2dd17a-f893-45b5-9189-5c1f3fa29c25.png)


This bug is caused by `__experimentalGetSpacingClassesAndStyles`

Aljullu commented:
> As long as our custom hook is feature-gated and no fatal error throws on WP 5.8, I think we can just focus on 5.9, it's not worth the effort to support 5.8 given that most of the global style functionalities are available from 5.9. cc @Aljullu for more insights : ).

Agree with this. :slightly_smiling_face: 

> This bug is caused by `__experimentalGetSpacingClassesAndStyles`

Would it be possible to check if `__experimentalGetSpacingClassesAndStyles` exists at [this point](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5515/files#diff-50fcc3107b21b78495f2049855c06d3f7624590b712a9174078cbd90004b70e9R37), and return early if it doesn't?

Then, we might probably need to do the same check [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5515/files#diff-bc6b772c8d9baf82eae8ac09948b4e392e5b526b1d9550c96d18d493a9e17738R30), so controls don't appear in the editor if the function doesn't exist.

gigitux commented:
> Would it be possible to check if `__experimentalGetSpacingClassesAndStyles` exists at [this point](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5515/files#diff-50fcc3107b21b78495f2049855c06d3f7624590b712a9174078cbd90004b70e9R37), and return early if it doesn't?
> 
> Then, we might probably need to do the same check [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5515/files#diff-bc6b772c8d9baf82eae8ac09948b4e392e5b526b1d9550c96d18d493a9e17738R30), so controls don't appear in the editor if the function doesn't exist.

Yes! Sure! I will add this check! Thanks for the feedback!

gigitux committed:
fix crash on WP 5.8

gigitux commented:
@Aljullu @dinhtungdu I just added a utility function [18642e6](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5515/commits/18642e60becfec59e289471ccae74275e6bc3de9) to check if there is support for the spacing style support.

Let me know, what do you think!

dinhtungdu commented on assets/js/atomic/blocks/product-elements/title/index.ts:
```diff
@@ -36,15 +37,16 @@ const blockConfig: BlockConfiguration = {
-			},
-		} ),
+		...( isFeaturePluginBuild() &&
+			hasSpacingStyleSupport() && {
```
@gigitux Isn't that check should be used for spacing below?

gigitux replied:
Update! Thanks for the double-check!

dinhtungdu commented on assets/js/atomic/blocks/product-elements/title/index.ts:
```diff
@@ -24,17 +26,34 @@ const blockConfig: BlockConfiguration = {
+				fontSize: true,
+				lineHeight: true,
+				__experimentalFontWeight: true,
+				__experimentalTextTransform: true,
+				__experimentalFontFamily: true,
+			},
+		} ),
+		...( isFeaturePluginBuild() && {
+			color: {
+				text: true,
+				background: true,
+				link: false,
+				gradients: true,
+				__experimentalSkipSerialization: true,
+			},
+		} ),
+		...( isFeaturePluginBuild() &&
+			hasSpacingStyleSupport() && {
+				spacing: {
+					margin: true,
+					__experimentalSkipSerialization: true,
 				},
-				typography: {
-					fontSize: true,
-				},
-		  }
-		: sharedConfig.supports,
+			} ),
```
Did `html: false` get removed on purpose? 

By the way, this part can be rewritten to:

```
		...sharedConfig.supports,
		...( isFeaturePluginBuild() && {
			typography: {
				fontSize: true,
				lineHeight: true,
				__experimentalFontWeight: true,
				__experimentalTextTransform: true,
				__experimentalFontFamily: true,
			},
			color: {
				text: true,
				background: true,
				link: false,
				gradients: true,
				__experimentalSkipSerialization: true,
			},
			...( hasSpacingStyleSupport() && {
				spacing: {
					margin: true,
					__experimentalSkipSerialization: true,
				},
			} ),
		} ),
```

dinhtungdu reviewed (commented):
@gigitux In my test with the latest `trunk` of Gutenberg, WP 5.9 beta 4, and WC 6.1.0-dev, the text color works for block style but doesn't work for global style.

https://user-images.githubusercontent.com/5423135/148975227-d0362ea4-8160-4606-b969-40bc3d384abb.mov

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/4965-product-title

gigitux commented:
> @gigitux In my test with the latest `trunk` of Gutenberg, WP 5.9 beta 4, and WC 6.1.0-dev, the text color works for block style but doesn't work for global style.
> 
>  Screen.Recording.2022-01-11.at.22.45.42.mov

Thanks for your feedback. If you save the custom color and you check the frontend, you will be able to see the custom color.

The problem, in this case, is that Gutenberg injects custom style for all anchor elements on global style
![image](https://user-images.githubusercontent.com/4463174/148999663-4532115c-ff8b-417e-aad3-3c0e1ba536d3.png)

We are not sure that this element is an `anchor` element because the user can enable or disable the link feature: if we want to fix the problem, we should enable both `link` and `text` color. To be honest, I think that it is not a great UX for the user, but from what I understand in this weeks there isn't an alternative.

dinhtungdu commented:
> The problem, in this case, is that Gutenberg injects custom style for all anchor elements on global style

@gigitux I opened #5548 to fix this issue by overriding the style for the anchor of the title in the editor. Please take a look!

> We are not sure that this element is an `anchor` element because the user can enable or disable the link feature: if we want to fix the problem, we should enable both `link` and `text` color. To be honest, I think that it is not a great UX for the user, but from what I understand in this weeks there isn't an alternative.

I think the title should use the same color setting regardless it's a link or not. Enable `link` color makes the block more complex.

dinhtungdu committed:
fix global style title color (#5548)

gigitux commented:
> I opened #5548 to fix this issue by overriding the style for the anchor of the title in the editor. Please take a look!

Thanks for the help! I didn't know that the `editor.scss` is loaded on the FSE editor page too. Thanks for the help!

Let me know if I can merge this PR!

dinhtungdu reviewed (approved):
:shipit:
